Fix diff function

ZachHaber opened the pull request:
Fixes #14
The updated diff function was incorrect
It's now correct and works with edges with IDs and edges without ids.
Note: without IDs, every time it changes, it removes the old and adds the new.

Fix issues with mutating the props!

ZachHaber commented on src/index.tsx:
```diff
@@ -140,22 +142,35 @@ const VisGraph = forwardRef<
+    if (updated.length) {
+      nodes.update(shallowClone(updated));
+    }
+    prevNodes.current = nodes.get();
```
I need to keep this as it was (same with prevEdges) which will work better with the isEqual calls. And I can just run `nodes.get` through the `diff` function.

Though I'm thinking it might make sense to remove the isEqual calls anyway... Since theoretically as long as the user doesn't generate the nodes/edges on the fly, it'll have changes, and if they do, as long as they have IDs, we'll more or less be doing the same thing. If they don't have IDs, then we'd be removing/adding them every time

WikKam commented:
I will run it tomorrow, I have some concerns about a few things but I don't know if we wanna adress them in this mr.

WikKam commented:
I will run it tomorrow, I have some concerns about a few things but I don't know if we wanna adress them in this mr.

- IMO tracking the previous values of nodes and edges should be at least done in seperate useEffect, the best solution would be to abstract pretty simple `usePreviousValue` hook.
- Two useEffects relating tracking the updated, added, deleted nodes basically do the same thing and should also be abstracted into some other custom hook

Other than that, in terms of code it is fine for me

WikKam commented on src/index.tsx:
```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
+    }
+    // Shallow clone edges to ensure props aren't mutated!
+    if (added.length) {
+      edges.add(shallowClone(added));
```
I would just
```
edges.add([...added])
```

WikKam commented on src/index.tsx:
```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
-    edges.update(updated);
+    const { added, removed, updated } = diff(edges.get(), graph.edges);
+    if (removed.length) {
+      edges.remove(removed);
```
I would just
```
edges.remove([...removed])
```

WikKam replied:
And correct me if I'm wrong but the if statement seems also unnecessarry, will something happen if we call `add, update, remove` with empty array?

ZachHaber replied:
Nothing happens if we call them with an empty array, so yeah, I can remove those if statements.
No need for `edges.remove([...removed])` as the remove function doesn't appear to mutate the inputs, plus if it's going in removed, it by definition came from the mutable `edges` source.

WikKam replied:
But do we need that `shallowClone` function to copy array or is `[...array]` enough for it?

ZachHaber replied:
For everything but the remove function, yes. we need the shallowClone, because it mutates the array element objects by adding an `id` onto it if there isn't one, which goes very much against how react works, so in order to protect the users of this component, it's best to shallow clone the individual elements of the array.

WikKam commented on src/index.tsx:
```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
+    }
+    // Shallow clone edges to ensure props aren't mutated!
+    if (added.length) {
+      edges.add(shallowClone(added));
```
same as above

ZachHaber replied:
The main reason I added these if checks was to avoid a bunch of no-op function calls since they were easily avoidable.

WikKam commented on src/index.tsx:
```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
+      edges.add(shallowClone(added));
+    }
+    if (updated.length) {
+      edges.update(shallowClone(updated));
```
same as above

WikKam commented on src/index.tsx:
```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
-    nodes.remove(removed);
-    nodes.add(added);
-    nodes.update(updated);
+    if (removed.length) {
+      nodes.remove(removed);
+    }
+    // Shallow clone nodes to ensure props aren't mutated!
+    if (added.length) {
+      nodes.add(shallowClone(added));
+    }
+    if (updated.length) {
+      nodes.update(shallowClone(updated));
+    }
```
Same as analogical section in the update edges useEffect

ZachHaber replied:
> I will run it tomorrow, I have some concerns about a few things but I don't know if we wanna adress them in this mr.
> 
> * IMO tracking the previous values of nodes and edges should be at least done in seperate useEffect, the best solution would be to abstract pretty simple `usePreviousValue` hook.
> * Two useEffects relating tracking the updated, added, deleted nodes basically do the same thing and should also be abstracted into some other custom hook
> 
> Other than that, in terms of code it is fine for me

I do see your point about setting up a custom hook to do the functionality that the edges/nodes need, I'll probably tie in the refs and other bits into that custom hook.

As for tracking the previous values of the nodes and edges, given that we are only using them here, we'd essentially be using an effect hook with the same dependencies to do that tracking, and given how small that functionality is, it's a bit wasteful in this specialized use case, especially since I need the ref to exist before this effect, but change the value of it afterwards, which unfortunately makes a `usePreviousValue` hook a bit more complicated to set up.

ZachHaber commented:
> I will run it tomorrow, I have some concerns about a few things but I don't know if we wanna adress them in this mr.
> 
> * IMO tracking the previous values of nodes and edges should be at least done in seperate useEffect, the best solution would be to abstract pretty simple `usePreviousValue` hook.
> * Two useEffects relating tracking the updated, added, deleted nodes basically do the same thing and should also be abstracted into some other custom hook
> 
> Other than that, in terms of code it is fine for me

I do see your point about setting up a custom hook to do the functionality that the edges/nodes need, I'll probably tie in the refs and other bits into that custom hook.

As for tracking the previous values of the nodes and edges, given that we are only using them here, we'd essentially be using an effect hook with the same dependencies to do that tracking, and given how small that functionality is, it's a bit wasteful in this specialized use case, especially since I need the ref to exist before this effect, but change the value of it afterwards, which unfortunately makes a `usePreviousValue` hook a bit more complicated to set up.

WikKam commented:
> > I will run it tomorrow, I have some concerns about a few things but I don't know if we wanna adress them in this mr.
> > 
> > * IMO tracking the previous values of nodes and edges should be at least done in seperate useEffect, the best solution would be to abstract pretty simple `usePreviousValue` hook.
> > * Two useEffects relating tracking the updated, added, deleted nodes basically do the same thing and should also be abstracted into some other custom hook
> > 
> > Other than that, in terms of code it is fine for me
> 
> I do see your point about setting up a custom hook to do the functionality that the edges/nodes need, I'll probably tie in the refs and other bits into that custom hook.
> 
> As for tracking the previous values of the nodes and edges, given that we are only using them here, we'd essentially be using an effect hook with the same dependencies to do that tracking, and given how small that functionality is, it's a bit wasteful in this specialized use case, especially since I need the ref to exist before this effect, but change the value of it afterwards, which unfortunately makes a `usePreviousValue` hook a bit more complicated to set up.

Sure, I am fine with that, lets do it in other MR. I think you are also right about the previous value, since the effects always maintain the same order

ZachHaber commented:
I went ahead and made a hook for it, because it was pretty straight forward.
As a side note, I just wanted to say that "MR" confused me at first as I'm used to calling them PRs. I blame localizations 😛
